external-crates/move/crates/move-vm-runtime/src/execution/interpreter/locals.rs

```diff
@@ -0,0 +1,305 @@
+/// is freed, we need to check that we are freeing the one on the end of the heap.
+#[derive(Debug)]
+pub struct StackFrame {
+    slice: Vec<Value>,
+}
+
+// -------------------------------------------------------------------------------------------------
+// Base (Machine-External) Heap
+// -------------------------------------------------------------------------------------------------
+
+impl Default for BaseHeap {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl BaseHeap {
+    pub fn new() -> Self {
+        Self {
+            values: HashMap::new(),
+            next_id: 0,
+        }
+    }
+
+    /// Allocate a slot for the value in the base heap, and then
+    pub fn allocate_and_borrow_loc(
+        &mut self,
+        value: Value,
+    ) -> PartialVMResult<(BaseHeapId, Value)> {
+        let next_id = BaseHeapId(self.next_id);
+        self.next_id += 1;
+        self.values.insert(next_id, Box::new(value));
+        let ref_ = self.borrow_loc(next_id)?;
+        Ok((next_id, ref_))
+    }
+
+    /// Moves a location out of memory, swapping it with `ValueImpl::Invalid`
+    pub fn take_loc(&mut self, ndx: BaseHeapId) -> PartialVMResult<Value> {
+        if self.is_invalid(ndx)? {
+            return Err(
+                PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                    .with_message("Cannot move from an invalid memory location".to_string()),
+            );
+        }
+
+        let Some(value_box) = self.values.get_mut(&ndx) else {
+            return Err(
+                PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                    .with_message(format!("Invalid index: {}", ndx)),
+            );
+        };
+
+        let value = std::mem::replace(value_box.as_mut(), Value::invalid());
+        Ok(value)
+    }
+
+    /// Borrows the specified location
+    pub fn borrow_loc(&self, ndx: BaseHeapId) -> PartialVMResult<Value> {
+        self.values
+            .get(&ndx)
+            .ok_or_else(|| {
+                PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR)
+                    .with_message(format!("Local index out of bounds: {}", ndx))
+            })
+            .and_then(|value| value.take_ref())
+    }
+
+    /// Checks if the value at the location is invalid
+    pub fn is_invalid(&self, ndx: BaseHeapId) -> PartialVMResult<bool> {
+        self.values
+            .get(&ndx)
+            .map(|value| matches!(value.as_ref(), &Value(ValueImpl::Invalid)))
+            .ok_or_else(|| {
+                PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                    .with_message(format!("Invalid index: {}", ndx))
+            })
+    }
+}
+
+// -------------------------------------------------------------------------------------------------
+// Machine (Runtime) Heap
+// -------------------------------------------------------------------------------------------------
+
+impl Default for MachineHeap {
```

Comment:
I. Hate. Default. I know that clippy will yell at you, but aghhhhhh

<img src="https://media1.giphy.com/media/ZebTmyvw85gnm/giphy.gif"/>